### PR TITLE
Fix retry on context deadline exceeded

### PR DESCRIPTION
**Pull request by vladaionescu**
This does two things:

* Does not retry on context deadline exceeded
* Does not unwrap context canceled and context deadline exceeded before returning the error from cloud client

### CODE REVIEW

File: cloud/client.go [hunk 1]
```diff
@@ -233,13 +239,16 @@ func shouldRetry(status int, body string, err error, warnFunc func(string, ...in
 		}
 		return true
 	}
-	if err != nil {
-		if errors.Cause(err) == ErrNoAuthorizedPublicKeys ||
-			errors.Cause(err) == ErrNoSSHAgent ||
-			strings.Contains(err.Error(), "failed to connect to ssh-agent") {
+	if callErr != nil {
+		causeErr := errors.Cause(callErr)
```
Comment (mikejholly) [inline cloud/client.go:243]: Can we use `errors.Is` for these?

File: cloud/client.go [hunk 2]
```diff
@@ -183,27 +183,33 @@ func (c *client) doCall(ctx context.Context, method, url string, opts ...request
...
 			// Some operations can be canceled gracefully, so we can signal this to higher-level functions.
 			// Note the actual error caught here is not an instance of context.Canceled,
 			// but is an error message that contains the HTTP call plus the string "context canceled".
-			return status, body, context.Canceled
-		}
-
-		if !shouldRetry(status, body, err, c.warnFunc) {
-			return status, body, err
+			switch {
+			case strings.Contains(callErr.Error(), "context canceled"):
```
Comment (mikejholly) [inline cloud/client.go:198]: I'm not quite following this. Are we getting `context canceled` as part of the API's response? If the HTTP client is hitting the cancellation, it should produce a legitimate `context.Canceled`, if I'm not mistaken.

**Comment by alexcb** [comment]
An integration test for this would be helpful; this could be done via:

```
earthly --server=http://127.0.0.1:1 +this-doesnt-exist-but-should-attempt-to-send-analytics
```

Currently (without this PR); this produces:

```
 1. Init 🚀
————————————————————————————————————————————————————————————————————————————————

           buildkitd | Found buildkit daemon as docker container (earthly-buildkitd)
           buildkitd | Updated image available. Restarting buildkit daemon...
           buildkitd | ...Done


 2. Build 🔧
————————————————————————————————————————————————————————————————————————————————

retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
failed to upload log: Post "http://127.0.0.1:1/api/v0/logs": dial tcp 127.0.0.1:1: connect: connection refused
Error: build target: build main: failed to solve: Earthfile line 0:0 target this-doesnt-exist-but-should-attempt-to-send-analytics not found
in		github.com/earthly/earthly:acb/builder-cleanup+this-doesnt-exist-but-should-attempt-to-send-analytics --DOCKERHUB_AUTH=true --DOCKERHUB_MIRROR=registry-1.docker.io.mirror.corp.earthly.dev --DOCKERHUB_TOKEN_SECRET=+secrets/earthly-technologies/dockerhub-mirror/pass --DOCKERHUB_USER_SECRET=+secrets/earthly-technologies/dockerhub-mirror/user --KEEPIT=secret
retrying http request due to unexpected error Put "http://127.0.0.1:1/analytics": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Put "http://127.0.0.1:1/analytics": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Put "http://127.0.0.1:1/analytics": dial tcp 127.0.0.1:1: connect: connection refused
retrying http request due to unexpected error Put "http://127.0.0.1:1/analytics": dial tcp 127.0.0.1:1: connect: connection refused
```

File: cloud/client.go (hunk 2 above)
Comment (vladaionescu) [reply cloud/client.go:198]: I believe that @alexcb might know the history behind looking for the string itself.

Comment (brandonSc) [reply cloud/client.go:198]: I added the string check and left a comment above. Apparently, I wasn't getting an instance of `context.Canceled` but a an error containing that in the string. That's why I caught it here and converted to an actual `context.Canceled`

Comment (brandonSc) [reply cloud/client.go:198]: As to why that is the case, I don't think I ever figured that out

File: cloud/client.go (hunk 1 above)
Comment (brandonSc) [reply cloud/client.go:243]: +1 - apparently this was [added semi-recently in pkg/errors](https://github.com/pkg/errors/releases/tag/v0.9.0) for compatibility with go1.13+ error

**Comment by brandonSc** [comment]
Nit: one difference in this new code seems to be that that the `context canceled` error gets printed now when you Ctrl-C in satellite launch or rm.
Screenshot 2022-08-18 142101

**Comment by vladaionescu** [comment]
Thanks for catching that @brandonSc - I'll adjust accordingly.

File: cloud/client.go [hunk 3]
```diff
@@ -183,27 +183,33 @@ func (c *client) doCall(ctx context.Context, method, url string, opts ...request
...
 	var body string
-	var err error
+	var callErr error
 	duration := time.Millisecond * 100
 	for attempt := 0; attempt < maxAttempt; attempt++ {
-		status, body, err = c.doCallImp(ctx, r, method, url, opts...)
-
-		if err != nil && strings.Contains(err.Error(), "context canceled") {
+		status, body, callErr = c.doCallImp(ctx, r, method, url, opts...)
+		if !shouldRetry(status, body, callErr, c.warnFunc) {
```
Comment (brandonSc) [inline cloud/client.go:190]: I think by moving this call further up, we are printing the warning "retrying HTTP request" when the error is context cancelled but not actually an instance of it according to the  `errors.Cause` check that was added in `shouldRetry`.  I pasted a screenshot of this in a different comment

**Review by alexcb — COMMENTED** [review]
reiterating my previous comment regarding a test for this (apparently github doesn't see that comment as me responding to this pr-feedback-request event)

**Comment by vladaionescu** [comment]
@alexcb, regarding the integration test: that wouldn't work for what this PR is fixing. The output is the same both before and after. It is correct for Earthly to retry on connection refused for both logs and analytics. The only difference though is that after the analytics timeout passes, in the new version it stops trying (`context.DeadlineExceeded` error now causes the retry to stop) and thus no additional output is shown after that.

**Review by brandonSc — APPROVED** [review]
🥇

**Comment by alexcb** [comment]
> that wouldn't work for what this PR is fixing. The output is the same both before and after.

It would probably need a more sophisticated stub-server, which seems like more work than is warranted for this PR then.

**Review by alexcb — APPROVED** [review]

**Pull request merged**